### PR TITLE
Add support for map in the runtime.

### PR DESCRIPTION
**Pull request by JakeWharton**
Refs #279.

### CODE REVIEW

File: wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java [hunk 1]
```diff
@@ -110,51 +111,63 @@ private void emitUint64(Long value, JsonWriter out) throws IOException {
...
-    return gson.fromJson(in, JsonElement.class);
-  }
-
-  private Object parseValue(Label label, Type valueType, JsonElement valueElement) {
-    if (label.isRepeated()) {
-      if (valueElement.isJsonNull()) {
+  private Object parseValue(FieldBinding<?, ?> fieldBinding, JsonElement element) {
+    if (fieldBinding.label.isRepeated()) {
+      if (element.isJsonNull()) {
         return Collections.emptyList();
```
Comment (swankjesse) [inline wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java]: off topic: this is sad.

File: wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 2]
```diff
@@ -219,10 +219,17 @@ public TypeName typeName(ProtoType protoType) {
   }
 
   private CodeBlock singleAdapterFor(Field field) {
-    ProtoType type = field.type();
+    return field.type().isMap()
+        ? CodeBlock.of("$N", field.name())
```
Comment (swankjesse) [inline wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: This is curious. I wonder how this works...

File: wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 3]
```diff
@@ -787,7 +816,13 @@ private String fieldName(ProtoType type, Field field) {
   }
 
   private TypeName fieldType(Field field) {
-    TypeName messageType = typeName(field.type());
+    ProtoType type = field.type();
+    if (type.isMap()) {
+      return ParameterizedTypeName.get(ClassName.get(Map.class), //
```
Comment (swankjesse) [inline wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: is `//` still useful? I nolove `//`

File: wire-runtime/src/main/java/com/squareup/wire/FieldBinding.java [hunk 4]
```diff
@@ -45,48 +46,67 @@ private static Method getBuilderMethod(Class<?> builderType, String name, Class<
...
+    this.keyAdapterString = wireField.keyAdapter();
     this.adapterString = wireField.adapter();
     this.redacted = wireField.redacted();
     this.messageField = messageField;
     this.builderField = getBuilderField(builderType, name);
     this.builderMethod = getBuilderMethod(builderType, name, messageField.getType());
   }
 
+  boolean isMap() {
+    return !keyAdapterString.isEmpty();
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/FieldBinding.java]: empty instead of null? sigh

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/FieldBinding.java]: I’m guessing this is a consequence of annotations. . .

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java [hunk 5]
```diff
@@ -66,6 +67,12 @@ public ProtoAdapter(FieldEncoding fieldEncoding, Class<?> javaType) {
     return new RuntimeEnumAdapter<>(type);
   }
 
+  /** Creates a new proto adapter for a map using {@code keyAdapter} and {@code valueAdapter}. */
+  public static <K, V> ProtoAdapter<Map<K, V>> createMapAdapter(ProtoAdapter<K> keyAdapter,
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: weirdly inconsistent for the line above `newEnumAdapter` and this `createMapAdapter` . . . maybe `newMapAdapter` ?

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java [hunk 6]
```diff
@@ -485,4 +492,84 @@ public String toString(E value) {
...
+    }
+
+    @Override public void encodeWithTag(ProtoWriter writer, int tag, Map<K, V> value)
+        throws IOException {
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        entryAdapter.encodeWithTag(writer, tag, entry);
+      }
+    }
+
+    @Override public Map<K, V> decode(ProtoReader reader) throws IOException {
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Maybe add a comment that this returns a single element map? That’s potentially surprising

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java [hunk 7]
```diff
@@ -485,4 +492,84 @@ public String toString(E value) {
...
+          case 1: key = entryAdapter.keyAdapter.decode(reader); break;
+          case 2: value = entryAdapter.valueAdapter.decode(reader); break;
+          default: // Ignore unknown tags in map entries.
+        }
+      }
+      reader.endMessage(token);
+
+      if (key == null) throw new IllegalStateException("Map entry with null key");
+      if (value == null) throw new IllegalStateException("Map entry with null value");
+      return Collections.singletonMap(key, value);
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Not sure about this call. I don’t think we ever do singleton lists do we?

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Would it be possible for this to return just a map entry?

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Or is this supposed to build a fullsize map and just doing a bad job?

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java [hunk 8]
```diff
@@ -485,4 +492,84 @@ public String toString(E value) {
...
+    }
+
+    @Override public Map<K, V> redact(Map<K, V> value) {
+      return Collections.emptyMap();
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Alternately, a map where we retain the keys and have redacted values?

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: (though that doesn’t work if it’s the keys we’re redacting)

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java [hunk 9]
```diff
@@ -167,7 +167,10 @@ B newBuilder() {
       FieldBinding<M, B> fieldBinding = fieldBindings.get(tag);
       try {
         if (fieldBinding != null) {
-          Object value = fieldBinding.singleAdapter().decode(reader);
+          ProtoAdapter<?> adapter = fieldBinding.isMap()
+              ? fieldBinding.adapter()
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: This is a bit surprising. I would have expected maps and lists to be symmetric.

File: wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
+package com.squareup.wire.internal;
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java]: nit: copyright

File: wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Override public Set<K> keySet() {
+    return map.keySet();
+  }
+
+  @Override public Collection<V> values() {
+    return map.values();
+  }
+
+  @Override public Set<Entry<K, V>> entrySet() {
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java]: Sadly the entries are mutable

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java]: You want `http://docs.oracle.com/javase/7/docs/api/java/util/AbstractMap.SimpleImmutableEntry.html`

File: wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  @Override public boolean containsValue(Object value) {
+    return map.containsValue(value);
+  }
+
+  @Override public V get(Object key) {
+    return map.get(key);
+  }
+
+  @Override public Set<K> keySet() {
+    return map.keySet();
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java]: Sadly mutable

File: wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  @Override public Collection<V> values() {
+    return map.values();
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/ImmutableMap.java]: Sadly mutable.

File: wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java [hunk 14]
```diff
@@ -53,12 +67,37 @@ private Internal() {
...
+    return result;
+  }
+
   public static <T> void redactElements(List<T> list, ProtoAdapter<T> adapter) {
     for (int i = 0, count = list.size(); i < count; i++) {
       list.set(i, adapter.redact(list.get(i)));
     }
   }
 
+  public static <T> void redactElements(Map<?, T> map, ProtoAdapter<T> adapter) {
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java]: Cool

File: wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java [hunk 15]
```diff
@@ -53,12 +67,37 @@ private Internal() {
...
+    for (Map.Entry<?, T> entry : map.entrySet()) {
+      entry.setValue(adapter.redact(entry.getValue()));
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java]: This can be mutating an immutable map?

Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java]: (upon inspection, yes, and working as intended)

File: wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java [hunk 16]
```diff
@@ -94,6 +133,17 @@ public static void checkElementsNotNull(List<?> list) {
     }
   }
 
+  /** Throw {@link NullPointerException} if {@code map} or one of its keys or values are null. */
+  public static void checkElementsNotNull(Map<?, ?> map) {
+    if (map == null) throw new NullPointerException("map == null");
+    if (map.containsKey(null)) throw new NullPointerException("map.containsKey(null)");
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java]: probably simpler to just check in the loop?  You’re already going through entry-by-entry?

File: wire-runtime/src/main/java/com/squareup/wire/internal/MutableOnWriteMap.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  private final Map<K, V> immutableMap;
+  Map<K, V> mutableMap;
+
+  MutableOnWriteMap(Map<K, V> immutableMap) {
+    this.immutableMap = immutableMap;
+    this.mutableMap = immutableMap;
+  }
+
+  @Override public Set<Entry<K, V>> entrySet() {
+    return mutableMap.entrySet();
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/internal/MutableOnWriteMap.java]: These entries are immutable, and mutating them will crash

File: wire-runtime/src/test/java/com/squareup/wire/MapTest.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Parameterized.class)
+public final class MapTest {
+  private static final ByteString BYTES = ByteString.decodeHex(
+      "0a0c0a036f6e6512050a034f6e650a0c0a0374776f12050a0354776f0a100a05746872656512070a055468726565");
```
Comment (swankjesse) [inline wire-runtime/src/test/java/com/squareup/wire/MapTest.java]: nice.

File: wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+import okio.ByteString;
+
+public final class Mappy extends Message<Mappy, Mappy.Builder> {
+  public static final ProtoAdapter<Mappy> ADAPTER = new ProtoAdapter_Mappy();
+
+  private static final long serialVersionUID = 0L;
+
+  @WireField(
+      tag = 1,
+      keyAdapter = "com.squareup.wire.ProtoAdapter#STRING",
```
Comment (swankjesse) [inline wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java]: dig.

File: wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+  )
+  public final Map<String, Thing> things;
+
+  public Mappy(Map<String, Thing> things) {
+    this(things, ByteString.EMPTY);
+  }
+
+  public Mappy(Map<String, Thing> things, ByteString unknownFields) {
+    super(ADAPTER, unknownFields);
+    this.things = Internal.immutableCopyOf("things", things);
```
Comment (swankjesse) [inline wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java]: Dig.

File: wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    @Override
+    public Mappy build() {
+      return new Mappy(things, super.buildUnknownFields());
+    }
+  }
+
+  private static final class ProtoAdapter_Mappy extends ProtoAdapter<Mappy> {
+    private final ProtoAdapter<Map<String, Thing>> things = ProtoAdapter.createMapAdapter(ProtoAdapter.STRING, Thing.ADAPTER);
```
Comment (swankjesse) [inline wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java]: Ahhh, so `things` can be the field name here. Makes sense.

File: wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+      writer.writeBytes(value.unknownFields());
+    }
+
+    @Override
+    public Mappy decode(ProtoReader reader) throws IOException {
+      Builder builder = new Builder();
+      long token = reader.beginMessage();
+      for (int tag; (tag = reader.nextTag()) != -1;) {
+        switch (tag) {
+          case 1: builder.things.putAll(things.decode(reader)); break;
```
Comment (swankjesse) [inline wire-runtime/src/test/proto-java/com/squareup/wire/map/Mappy.java]: ... and this is the code that decodes a single entry. Still tempted to make `things` a `ProtoAdapter<Map.Entry<String, Thing>>`

File: wire-runtime/src/test/proto/map.proto [hunk 23]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto2";
+
+package com.squareup.wire.map;
+
+message Mappy {
+  map<string, Thing> things = 1;
```
Comment (swankjesse) [inline wire-runtime/src/test/proto/map.proto]: sweet

**Comment by swankjesse** [comment]
# nice!

Most of my comments are around how horribly painful it is to implement an immutable map. ([shoutout to 2008](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ImmutableMap.java#L47)!)

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java (hunk 7 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: This is because of how the encoding works.

Think of how we do repeated messages: we create an empty list and append a single message at a time as we see them. This is because the proto adapter only knows how to decode the message but the parent adapter is the one that matches on the tag and delegates.

Maps are effectively a repeated message of key/value pairs. Unlike lists, however, there are two types that we need to decode along with the enclosing message. This means that we either can re-use the `Map<K, V>` adapter which is needed for encoding and getting the encoded size for reading single-element `Map<K, V>` instances, or we can create a separate adapter for just `Map.Entry<K, V>`.

If you compare the implementation of `Collections.singletonMap` with `SimpleMapEntry` they're effectively the same class. The difference is that `Map` has a `putAll(Map)` method where there is no `put(Map.Entry)`. This means that if we want the calling code to be a one-liner, the singleton map + `putAll` is the easiest way to go.

If we use an adapter for `Map.Entry` instead, you also have a problem of encoding. If you call `asRepeated()` which returns a `ProtoAdapter<List<Map.Entry<K, V>>>` we have no way to represent the list as that type without a copy of `.entrySet()`.

Simply put: neither `ProtoAdapter<Map<K, V>>` nor `ProtoAdapter<Map.Entry<K, V>>` ends up being ergonomic for all three methods, but I think `Map` is more ergonomic (and there's parity with `List`).

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java (hunk 8 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Yep. You either mark the field as redacted and get empty or we ask the values to redact themselves.

File: wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java (hunk 15 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/internal/Internal.java]: This is only called from generated code where we're assured it's a mutable map.

**Comment by JakeWharton** [comment]
Dropped the immutable/mutable-on-write map implementations for now. Updated and responded to everything else.

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java (hunk 9 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: In this case `singleAdapter()` delegates to call `ProtoAdatper.get` of `WireField.adapter` which in `Map`'s case is the adapter for the value type only.

I can look to make it more logical. The runtime adapter and field binding classes are a bit of a mess right now.

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java (hunk 6 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: I can add on `newMapAdapter` but it's also true for `asRepeated()` and `List`. Because of how adapters work we can only ever decode one element at a time. In the case of `List` we delegate directly to the item adapter and skip the singleton list. Unfortunately because of the complex type here we can't do that.

Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Here's the implementation for `asRepeated`:
```java
@Override public List<E> decode(ProtoReader reader) throws IOException {
  E value = ProtoAdapter.this.decode(reader);
  return Collections.singletonList(value);
}
```

File: wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java [hunk 24]
```diff
@@ -110,51 +111,63 @@ private void emitUint64(Long value, JsonWriter out) throws IOException {
...
+      return result;
     }
-  }
 
-  private Object readJson(Type valueType, JsonElement element) {
-    return gson.fromJson(element, valueType);
-  }
+    if (fieldBinding.isMap()) {
+      if (element.isJsonNull()) {
+        return Collections.emptyList();
```
Comment (swankjesse) [inline wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java]: emptyMap ?

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java [hunk 25]
```diff
@@ -66,6 +67,18 @@ public ProtoAdapter(FieldEncoding fieldEncoding, Class<?> javaType) {
     return new RuntimeEnumAdapter<>(type);
   }
 
+  /**
+   * Creates a new proto adapter for a map using {@code keyAdapter} and {@code valueAdapter}.
+   * <p>
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: For what it’s worth, putting `<p>` on its own line like this makes CMD+ALT+L work less well.

**Comment by swankjesse** [comment]
LGTM. Just one issue with emptyList().

File: wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java (hunk 24 above)
Comment (JakeWharton) [inline wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java]: Good catch. Adding tests.

File: wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java (hunk 25 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java]: Hmm it doesn't have any effect for me either way. It stays in the same location and the rendered output looks the same.
